File: kmeans.go
// Katie Christensen
// CSCI 512
// Program 2
// February 15, 2023

package main

import ( 
	"fmt"
	"os"
	"io"
	"strconv"
	"math"
	"strings"
)

var BIG_float64 = math.Inf(1)
var HYPHEN = "-"
var kdefault = 5
var km1default = 4
var InitialScoresArraySize = 100

var scores []float64
var bestAllk = BIG_float64
var k = kdefault
var km1 = km1default

type cluster struct {
	start int
	stop int
	mean float64
	tightness float64
}

var clusters []cluster
var bestPartition []cluster


func clustmean(i int) (float64) {
	var d float64
	var j, clustcount int 

	d = 0.0
	clustcount = (clusters[i].stop - clusters[i].start) + 1

	for j = clusters[i].start; j <= clusters[i].stop; j++ {
		d = d + scores[j]
	}

	return d/float64(clustcount)
}


func getstats() {
	var i, j int = 0,0
	var tmpd, tmpd2, tightnessAllk float64
	
	tightnessAllk = 0.0

	for i = 0; i < k; i++ {
		tmpd = clustmean(i)
		clusters[i].mean = tmpd
		clusters[i].tightness = 0.0
		tmpd2 = 0.0
		
		for j = clusters[i].start; j <= clusters[i].stop; j++ {
			tmpd2 = scores[j]-tmpd
			clusters[i].tightness += math.Abs(tmpd2)
		}
		
		tightnessAllk = tightnessAllk + clusters[i].tightness
	}

	if tightnessAllk < bestAllk {
		bestAllk = tightnessAllk
		
		for i = 0; i < k; i++ {
			bestPartition[i].start = clusters[i].start
	    	bestPartition[i].stop = clusters[i].stop
	    	bestPartition[i].mean = clusters[i].mean
	    	bestPartition[i].tightness = clusters[i].tightness 
		}
	}
}


func findbest(level int) {
	clusters[level].stop=clusters[level+1].start-1
	
	if level == 0 {
		getstats()
	} else {
		for clusters[level].start = level; clusters[level].start < clusters[level+1].start; clusters[level].start++ {
			findbest(level - 1)
		}
	}
}


func usage() {
	fmt.Println("Usage: ")
	fmt.Println("     [-k value] where value is the number of clusters to generate")
	fmt.Println("     [-h] print usage message ")
	fmt.Println("Defaults: k = 5")
	os.Exit(-1)
}


func kSizeError() {
	fmt.Println("Error: K  value must be a positive integer > 1")
	usage()
}


func commandOptError() {
	fmt.Println("Unrecognized option ")
	usage()
}


func main() {
	var i, index, optarg, j, n, nm1, CurrentScoresArraySize int = 0,0,0,0,0,0,0
	var c []string
	var score float64
	var flag string
	
	CurrentScoresArraySize = InitialScoresArraySize;
	
	scores = make([]float64, InitialScoresArraySize )

	if len(os.Args) == 1 {
		fmt.Println("Defaulting to 5 clusters")
	} else {
		for index = 1; index < len(os.Args); index++ {
			if strings.HasPrefix(os.Args[index], "-") && os.Args[index] != "--" && os.Args[index] != "-" {
				c = strings.Split(os.Args[index], "-")
				flag = c[1]
				switch flag {
				case "k":
					optarg, _ = strconv.Atoi(os.Args[index+1])
					k = optarg
					km1 = k - 1
					break
				case "h":
					usage()
				default:
					commandOptError()
				}
			} 
		}
	}

	if k <= 1 {
		kSizeError()
	}
	
	clusters = make([]cluster, k) 
	bestPartition = make([]cluster, k) 

	for {
		_, err := fmt.Scanf("%f", &score)
		
		if err == io.EOF {
			break
		} else {
			scores[i] = score
			i++
			if i > CurrentScoresArraySize {
				fmt.Printf("Number of points in the distribution is getting pretty large %d\n",i)
				fmt.Println("This is an exact solution to an NP-hard algorithm")
				fmt.Println("You may have to wait a while for results")
				
				CurrentScoresArraySize += 100
				
				scores = append(scores, float64(CurrentScoresArraySize)) 
				
				if scores == nil {
					fmt.Println("Bad news.\nrealloc failed trying to extend the scores array.\n Your data set must be insanely large???")
					os.Exit(-2)
				}
			}
		}
	}

	if k > i {
		fmt.Printf("Number of clusters requested %d is > the number of data points %d\n",k,i)
		fmt.Println("Terminating. Get your head awake and try again.")
		os.Exit(-3)
	} else {
		fmt.Printf("Number of points in the distribution is %d\n",i)
	}

	n = i
	nm1 = n - 1

	for i := 0; i < k; i++ {
		clusters[i].mean=scores[i]
      	clusters[i].start=i
      	clusters[i].stop=i
      	clusters[i].tightness=0.0
	}

	clusters[km1].stop = nm1
	
	for clusters[km1].start = km1; clusters[km1].start < n; clusters[km1].start++ {
		findbest(km1-1);
	}

	for i := 0; i < k; i++ {
		fmt.Printf("Partition %d = %d to %d scores=", i, bestPartition[i].start, bestPartition[i].stop)
		
		for j = bestPartition[i].start; j <= bestPartition[i].stop; j++ {
			fmt.Printf("%.2f ", scores[j])
		}

		fmt.Printf("mean=%.2f ", bestPartition[i].mean)
		fmt.Printf("tightness=%.2f\n", bestPartition[i].tightness / float64((bestPartition[i].stop - bestPartition[i].start + 1)))
	}
}

File: kmeans.c
/* kmeans.c Provided by Dr. Meehan from Western Washington University
   
   The distribution fed to this program should be sorted
   It doesn't matter whether is is sorted ascending or descending

   The program does not check that the input is sorted.
   If you feed it unsorted data the "results" are meaningless

   ---

   This is an EXACT solution to an NP-hard problem
   Don't be scared. It will produce a solution very quickly
   for small data sets with a reasonable k. We are dealing with
   one dimensional data rather than the general kmeans problem
   with arbitrarily large dimensional vectors.
   In particular for application to a distributioon of  scores 
   in a class of typical size the results are instantaneous.
   I routinely use it with k = 10 on 40 point distributions 
*/

#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <unistd.h>

#define BIG_double (INFINITY)
#define HYPHEN '-'
#define kdefault 5
#define km1default 4
#define InitialScoresArraySize 100  // we grow the scores array
                                    // in increments of 100 as needed

static double *scores; 	        // the distribution to be analyzed
                                // assumed to be sorted if it isn't
                                // results are meaningless 
static double bestAllk=BIG_double;

static int k=kdefault;
static int km1=km1default;

typedef struct 
{ 
  int     start;
  int     stop;
  double  mean;
  double  tightness;
} cluster;

// number of clusters we want
static cluster *clusters; // k cluster structs where the partitions are varied
static cluster *bestPartition;   // tightest overall partition known

double clustmean(int i)         // compute mean for cluster i
{ double  d=0.0;
  int clustcount=(clusters[i].stop - clusters[i].start) + 1;
  int j;
  for (j=clusters[i].start; j<=clusters[i].stop; j++) d+=scores[j];
  return d/clustcount;
} // end clustmean

void getstats()
{ int i,j;
  double tmpd,tmpd2;
  double tightnessAllk=0.0;
  for (i=0; i<k; i++)
    { tmpd=clustmean(i);
      clusters[i].mean=tmpd; 
      clusters[i].tightness=0.0;
      tmpd2=0.0;
      for (j=clusters[i].start; j<=clusters[i].stop; j++)
	{ tmpd2=scores[j]-tmpd;
	  clusters[i].tightness+=fabs(tmpd2);
	}
      tightnessAllk += clusters[i].tightness;
    };
  if ( tightnessAllk < bestAllk) 
    {bestAllk=tightnessAllk; 
      for (i=0; i<k; i++)
	{ bestPartition[i].start=clusters[i].start;         // 
	  bestPartition[i].stop=clusters[i].stop;
	  bestPartition[i].mean=clusters[i].mean;
	  bestPartition[i].tightness=clusters[i].tightness;  
	}
    }
} // end getstats

void findbest(int level)
{
  clusters[level].stop=clusters[level+1].start-1;
  if (level==0) getstats();
  else
    {
      for (clusters[level].start=level;				\
	   clusters[level].start<clusters[level+1].start;	\
	   ++clusters[level].start)
	findbest(level-1);
    }
}

void usage()
{
  printf("Usage: \n");
  printf("     [-k value] where value is the number of clusters to generate\n");
  printf("     [-h] print usage message \n");
  printf("Defaults: k = 5\n");
  exit(-1);
};

void kSizeError()
{
  printf("Error: K  value must be a positive integer > 1\n");
  usage();
};

void commandOptError()
{
  printf("Unrecognized option \n");
  usage();
};

int main (int argc, char *argv[])
{ int i=0;
  int c,tmpint,j,n,nm1,p,q,r,s,t,level,CurrentScoresArraySize;
  double tmpmean=0.0;
  double score;
  char optstring[]="hk:";
  CurrentScoresArraySize=InitialScoresArraySize;
  scores = (double *) malloc(sizeof(double)*InitialScoresArraySize);
  /* get the command line parameters to find out what k value we will
     be using
  */
  if (argc==1)
    printf("Defaulting to 5 clusters\n");
  else
    while ((c = getopt(argc,argv,optstring)) != -1)
      switch (c)
	{
	case 'k':
	  k = atoi(optarg);
	  km1=k-1;
	  break;
	case 'h':
	  usage();
	default:
	  commandOptError();
	};

  if (k <=1 ) kSizeError();
   
  clusters = (cluster *) malloc(sizeof(cluster)*k);
  bestPartition =(cluster *)malloc(sizeof(cluster)*k);
  while (scanf("%lf", &score) != EOF)
    {
      scores[i]=score;
      i++;
      if (i > CurrentScoresArraySize)
	{
	  printf("Number of points in the distribution is getting pretty large %d\n",i);
	  printf("This is an exact solution to an NP-hard algorithm\n");
	  printf("You may have to wait a while for results\n");
	  CurrentScoresArraySize+=100;
	  scores=realloc(scores,sizeof(double)*CurrentScoresArraySize);
	  if (scores==NULL)
	    {
	      printf("Bad news.\nrealloc failed trying to extend the scores array.\n Your data set must be insanely large???\n");
	      exit(-2);
	    }
	}
    }
  if (k > i )
    {
      printf("Number of clusters requested %d is > the number of data points %d\n",k,i);
      printf("Terminating. Get your head awake and try again.\n");
      exit(-3);
    }
  else
    printf("Number of points in the distribution is %d\n",i);
  n=i;
  nm1=n-1;

  //set initial cluster means and positions
  for (i=0; i< k; i++)
    {
      clusters[i].mean=scores[i];
      clusters[i].start=i;
      clusters[i].stop=i;
      clusters[i].tightness=0.0;
    }

  clusters[km1].stop=nm1;
  for (clusters[km1].start=km1; clusters[km1].start<n;++clusters[km1].start)
    {
      findbest(km1-1);
    }
 
  for (i=0; i<k ; i++)
    { printf("Partition %d = %d to %d scores=",i,bestPartition[i].start,bestPartition[i].stop);
      for (j=bestPartition[i].start;j<=bestPartition[i].stop;j++)
	{ printf("%2.2lf ",scores[j]);
	}
      printf("mean=%2.2lf ",bestPartition[i].mean);
      printf("tightness=%2.2lf\n",bestPartition[i].tightness/
	     (bestPartition[i].stop-bestPartition[i].start+1));
    }
} // end program
